rdmo/projects/validators.py

```diff
@@ -56,3 +57,77 @@ def __call__(self, data, serializer):
                 raise serializers.ValidationError({
                     'quota': [_('The file quota for this project has been reached.')]
                 })
+
+
+class ValueTypeValidator:
+
+    requires_context = True
+
+    def __call__(self, data, serializer):
+        text = data.get('text')
+        value_type = data.get('value_type')
+
+        if text and settings.PROJECT_VALUES_VALIDATION:
```

Comment:
this type of pattern `if.. elif.. elif... etc..`  is really not readable or maintainable..
do you mind using a mapping and selecting the validators from there instead?

Maybe for inspiration (GPT generated):
```py
from django.core.validators import URLValidator, RegexValidator, EmailValidator
from django.core.exceptions import ValidationError
from django.utils.translation import gettext as _
from rest_framework import serializers
from datetime import datetime

def validate_text_based_on_value_type(text, value_type):
    if not text or not settings.PROJECT_VALUES_VALIDATION:
        return

    validation_map = {
        'url': {
            'enabled': settings.PROJECT_VALUES_VALIDATION_URL,
            'validator': URLValidator(),
            'message': lambda e: e.message,
        },
        'integer': {
            'enabled': settings.PROJECT_VALUES_VALIDATION_INTEGER,
            'validator': RegexValidator(settings.PROJECT_VALUES_VALIDATION_INTEGER_REGEX),
            'message': settings.PROJECT_VALUES_VALIDATION_INTEGER_MESSAGE,
        },
        'float': {
            'enabled': settings.PROJECT_VALUES_VALIDATION_FLOAT,
            'validator': RegexValidator(settings.PROJECT_VALUES_VALIDATION_FLOAT_REGEX),
            'message': settings.PROJECT_VALUES_VALIDATION_FLOAT_MESSAGE,
        },
        'boolean': {
            'enabled': settings.PROJECT_VALUES_VALIDATION_BOOLEAN,
            'validator': RegexValidator(settings.PROJECT_VALUES_VALIDATION_BOOLEAN_REGEX),
            'message': settings.PROJECT_VALUES_VALIDATION_BOOLEAN_MESSAGE,
        },
        'date': {
            'enabled': settings.PROJECT_VALUES_VALIDATION_DATE,
            'validator': RegexValidator(settings.PROJECT_VALUES_VALIDATION_DATE_REGEX),
            'message': settings.PROJECT_VALUES_VALIDATION_DATE_MESSAGE,
        },
        'datetime': {
            'enabled': settings.PROJECT_VALUES_VALIDATION_DATETIME,
            'validator': lambda value: datetime.fromisoformat(value),
            'message': _('Enter a valid datetime.'),
        },
        'email': {
            'enabled': settings.PROJECT_VALUES_VALIDATION_EMAIL,
            'validator': EmailValidator(),
            'message': lambda e: e.message,
        },
        'phone': {
            'enabled': settings.PROJECT_VALUES_VALIDATION_PHONE,
            'validator': RegexValidator(settings.PROJECT_VALUES_VALIDATION_PHONE_REGEX),
            'message': settings.PROJECT_VALUES_VALIDATION_PHONE_MESSAGE,
        },
    }

    if value_type in validation_map:
        config = validation_map[value_type]
        if config['enabled']:
            try:
                config['validator'](text)
            except ValidationError as e:
                message = config['message'](e) if callable(config['message']) else config['message']
                raise serializers.ValidationError({'text': [message]}) from e
            except ValueError as e:  # specifically for datetime.fromisoformat
                if value_type == 'datetime':
                    raise serializers.ValidationError({'text': [config['message']]}) from e
```

Reply:
no I don't think so, the code example has to many indirection without gaining much in simplicity. I added a function though, to make it more readable.